common/achievement/evaluate.ts

```diff
@@ -0,0 +1,114 @@
+
+    const eventsByMetric: Record<string, Achievement_event[]> = {};
+    for (const event of achievementEvents) {
+        if (!eventsByMetric[event.metric]) {
+            eventsByMetric[event.metric] = [];
+        }
+        eventsByMetric[event.metric].push(event);
+    }
+
+    const resultObject: Record<string, number> = {};
+    resultObject['recordValue'] = recordValue;
+
+    for (const key in dataAggregation) {
+        if (!dataAggregation[key]) {
+            continue;
+        }
+        const dataAggregationObject = dataAggregation[key];
+        const metricName = dataAggregationObject.metric;
+
+        const bucketCreator = dataAggregationObject.createBuckets || 'default';
+        const bucketAggregator = dataAggregationObject.bucketAggregator || 'count';
+
+        const aggregator = dataAggregationObject.aggregator;
+
+        const eventsForMetric = eventsByMetric[metricName];
+        if (!eventsForMetric) {
+            continue;
+        }
+        // we take the relation from the first event, that posesses one, in order to create buckets from it, if needed
+        const relation = eventsForMetric.find((event) => event.relation)?.relation;
+
+        const bucketCreatorFunction = bucketCreatorDefs[bucketCreator].function;
+        const bucketAggregatorFunction = aggregators[bucketAggregator].function;
+
+        const aggregatorFunction = aggregators[aggregator].function;
+
+        if (!bucketCreatorFunction || !bucketAggregatorFunction || !aggregatorFunction) {
+            logger.error(
+                `No bucket creator or aggregator function found for ${bucketCreator}, ${aggregator} or ${bucketAggregator} during the evaluation of achievement`
+            );
+            return;
+        }
+
+        let bucketContext: AchievementContextType;
+        if (relation) {
+            bucketContext = await getBucketContext(relation);
+        }
+
+        const buckets = bucketCreatorFunction({ recordValue: recordValue, context: bucketContext });
+
+        const bucketEvents = createBucketEvents(eventsForMetric, buckets);
+        const bucketAggr = bucketEvents.map((bucketEvent) => bucketAggregatorFunction(bucketEvent.events.map((event) => event.value)));
+
+        const value = aggregatorFunction(bucketAggr);
+        resultObject[key] = value;
+    }
+    // TODO: return true if the condition is empty (eg. a student finishes a course and automatically receives an achievement)
+    const evaluate = swan.parse(condition);
+    const value: boolean = await evaluate(resultObject);
+
+    return {
+        conditionIsMet: value,
+        resultObject,
+    };
+}
+
+export function createBucketEvents(events: Achievement_event[], bucketConfig: BucketConfig): BucketEvents[] {
+    switch (bucketConfig.bucketKind) {
+        case 'default':
+            return createDefaultBuckets(events, bucketConfig);
+        case 'time':
+            return createTimeBuckets(events, bucketConfig);
+    }
```

Comment:
please throw an error if none was found.

Reply:
The case could not be nothing, following the code before. But, if it was, we return the default bucket events now.

Reply:
My idea was that it would save us from typo's when adding new buckets in the future.
I just checked, and the typesystem would catch it. Good job 👍